src/_pytest/logging.py

```diff
@@ -615,77 +619,63 @@ def pytest_collection(self) -> Generator[None, None, None]:
             else:
                 yield
 
-    @contextmanager
-    def _runtest_for(self, item, when):
-        with self._runtest_for_main(item, when):
+    def _runtest_for(self, item: nodes.Item, when: str) -> Generator[None, None, None]:
+        """Implements the internals of pytest_runtest_xxx() hook."""
+        if self.log_cli_handler is not None:
+            self.log_cli_handler.set_when(when)
+
+        with catching_logs(
+            handler=self._runtest_log_capture_handler,
+            formatter=self.formatter,
+            level=self.log_level,
+        ) as log_handler:
+            log_handler.reset()
+            item._store[catch_log_handler_key] = log_handler
+            item._store[catch_log_records_key][when] = log_handler.records
+
             if self.log_file_handler is not None:
                 with catching_logs(self.log_file_handler, level=self.log_file_level):
                     yield
             else:
                 yield
```

Comment:
@bluetech I like the usage of `yield-from` in this PR! WDYT about converting this `if self.log_file_handler` block (and also the block in `pytest_collection`) also into a generator from which you yield? Unfortunately I can't come up with a good name for this generator. What about `yield from self.forward_logs_to_logfile()`?

Reply:
I think we can make it mostly go away! But if not, I'll try your suggestion.